### ISSUE TITLE
[design flaw] config as code is wrong

### ISSUE THREAD

**Issue by samuelstroschein** [inlang/inlang#1143]
## Problem 

The `js` config can't be modified by apps.

- users can't add languages in the editor 
- https://github.com/inlang/inlang/issues/1139
- the list goes on...

## Proposal 

Replace the JS config with a JSON config that can be modified by apps (similar to VSCode settings). 

- enables apps to be much more powerful 
- apps can work independently of developers (translator wants to add a plugin for example)
- only importing plugins is what user config do anyways, nobody uses the power of code now that plugins are available
- entails the requirement for #1140

```ts
{
  "referenceLanguageTag":"en",
  "languageTags":[
    "en",
    "de",
    "fr"
  ],
  "plugins":{
    "inlang.standardLintRules":{
      "module":"https://cdn.jsdelivr.net/npm/@inlang/plugin-standard-lint-rules@3/dist/index.js"
    },
    "inlang.plugin-json":{
      "module":"https://cdn.jsdelivr.net/npm/@inlang/plugin-json@3/dist/index.js",
      "options":{
        "pathPattern":"source-code/website/lang/{language}.json"
      }
    }
  },
  "lint":{
    "rules":{
      "inlang.missingMessage":"off"
    }
  }
}
```

**Comment by samuelstroschein** [comment]
For plugins we could also make "where you host your plugin" easy: Any JSDelivr link works. NPM, GitHub, we don't care. 

In the future, we likely need our own registry.

**Comment by samuelstroschein** [comment]
Also `languageTags` must always be provided. Plugins have an API that doesn't mix with config. For legacy purposes, plugins can have a `extendLanguageTags` callback that should phase out over time. 

We want app users to have the power to add languageTags and messages as they desire.

**Comment by NilsJacobsen** [comment]
- In the future, we should be able to change settings without editing code. There should be an option in the future to distinguish between private and project settings (user configs / app internal settings via localstorage / ...)
- JSON is easier to understand for devs
- Typesafty is easier

Agree with the proposal 👍

**Comment by ivanhofer** [comment]
I agree with the point that inlang does not need a json config file.

I took another look at the initial [RFC](https://github.com/inlang/inlang/blob/main/rfcs/core-architecture/RFC.md) and compared it to the RFC of [`pipeli18ne`](https://github.com/pipeli18ne/RFC/blob/main/01-link-layer.md) to see why we both came up with the "config as code" approach. For `pipeli18ne` it is beneficial because it is a more flexible system (pipelines), primarely designed for developers. But inlang uses "apps" to offer it's functionality. Those apps will contain the complexity.

The reason why we did introduce "plugins" is to make the inlang config simpler. Now it is so simple, that we don't need the config as code anymore 💪. Because the config needs to be parsed also inside a browser the config as code has also it's limitations, because it can't use standards like regular imports of relative files, which always made it odd to set one up (`$env`). So in my opinion a JS config without the full power of JS does not make that much sense.

Migrating to a `json`, or any other format (e.g. `yaml`) could fit. Just don't use `xml` as a format!! 😅

If we choose `json` as a format, we should use `JSON5` to support comments.
[Quote from @samuelstroschein](https://github.com/inlang/inlang/blob/main/rfcs/core-architecture/RFC.md#2-how-is-inlang-configured): _"Furthermore, JavaScript as config solves two common config file annoyances. First, comments are supported, and type annotations via JSDoc/TypeScript enable autocomplete and type safety."_

One key point in favor of `js` was the possibilty to offer typesafety (which fall apart due to the non-standard `$env` imports).
This is also an option for `json` and `yaml` files, but only in a generic way (we can't really type the dynamic plugins object). In the future we could use the IDE extension to show configuration issues inline. And for now we should make sure that each plugin checks if the passed config matches it's expectations.

I have some details which we should consider for the refactoring:
 - `languageTags` can be defined in the inlang config, but also a plugin should be able to define them. The important thing should be, that once the app is initialized, that prop should be here
 - the thing that we currently call "plugin" sohuld be renamed to "message-plugin". "lint-rules" are also "plugins", but have their dedicated naming. I would add a new prop "messages" to the config to make it clear that this "plugin" is responsible to read/write messages. Question? Is it allowed to have multiple "messages" plugins? Could make sense as I have worked with apps where part of the translations were in a DB and some parts in `.json` files. Then each `Message` would also need a new prop to know which plugin is responsible. But this is probably a use case we can work on in the future.
 - combine lint plugin definition and lint configuration. "lint-plugins" should not be treated as plugins. They should have their own spot in the config to make it easier to configure them.
 - it should be encouraged to reference plugins (and lint-rules) by their identifier + version and not the url. "@inlang/plugin-json@3" is all we need to fetch the code from the jsdelivr CDN. If something is hosted somewhere else (private Company server), it still can be referenced via an url.

Here is an example for a config:
 ```json
 {
  "referenceLanguageTag": "en",
  "messages": {
    "https://cdn.jsdelivr.net/npm/@inlang/plugin-json@3/dist/index.js": {
        "pathPattern":"source-code/website/lang/{language}.json"
    }
  },
  "lint":{
    "inlang.standardLintRules@3": {
      "missingMessage":"off"
    }
  }
}
```

Additional considerations:
 - each plugin should define it's version and the version of core it uses as a prop, which will make future migration easier because we can e.g. have some conditional logic inside core or an app that makes it possible to use older plugins for a few months, without requiring to coordinate everything.

**Comment by samuelstroschein** [comment]
I am spec'ing the APIs which will serve as foundational "let's discuss those APIs". Lot's of things will become more clear. 

---

> @ivanhofer If we choose json as a format, we should use JSON5 to support comments.

Unfortunately... JSON5 is not suitable :( JSON5 (or any other json with comments) does not support roundtrip parsing and serialization. Furthermore, every app would need to install a JSON5 parser which increases complexity. 

I know we love comments. But, I don't see a viable alternative to plain JSON :(

**Comment by samuelstroschein** [comment]
> Why do we need serialization?
The parser would be bundled inside core (which provides the logic to parse everything, setup plugins etc.), so this is not an issue. 

CRUD ops on the config. 

- adding plugins via apps
- changing languages via apps
- changing lint rule levels via apps
- ...

**Comment by NiklasBuchfink** [comment]
I agree with the proposed changes and the input presented by @ivanhofer.

Splitting plugins by purpose makes so much sense when trying to generate or automate this process. I still dream of this simple way to configure or setup inlang for each repository:

![Config create modal](https://github.com/inlang/inlang/assets/59048346/c3380ab8-c56a-4c8e-8c28-fde1083fd960)

**Comment by felixhaeberle** [comment]
Madame et Monsieur – hold on to the handrail, @NiklasBuchfink is sketching on a Post-It again 😄

(context: while in our bachelors, niklas often summarized/sketched good ideas on one single post-it while other people were trying to build up big sketches on the whiteboard 😅)

**Comment by ivanhofer** [comment]
> CRUD ops on the config.

I still don't get why `JSON5` won't work here

**Comment by felixhaeberle** [comment]
@ivanhofer If I remember correctly (I do not have experience working with json5 – please correct me if I'm wrong), it does not have built-in roundtrip support in the sense that it is not directly designed to preserve all aspects of the original JSON5 data structure during parsing and serialization.

**Comment by samuelstroschein** [comment]
@ivanhofer JSON5 doesn't have roundtrip capabilities. 

- you can't serialize a json5 while preserving comments/"the features" of JSON5
- vscode had a lengthy debate, and I quote a maintainer "JSON5 is a toy project that doesn't work in real use cases" or something along those lines 
- https://github.com/OWASP/Maturity-Models/issues/61
- https://github.com/json5/json5/issues/121#issuecomment-223326811

**Comment by ivanhofer** [comment]
Ah, I did not know that 🙈. I was assuming everything get's preserved. We will have the same formatting issues with json. So maybe use a library that works on an AST? https://github.com/rse/json-asty or https://github.com/noahsug/json5-writer

**Comment by samuelstroschein** [comment]
Please vote on JSON as the config format. 

**Pros**

- roundtrip capabilities
- format with a native parser (in nearly every programming language) `JSON.parse`
- "everyone" understands JSON compared to TOML for example
- JSON just works (compared to YAML)

**Cons**

- no native comments

---

> @ivanhofer  We will have the same formatting issues with json. So maybe use a library that works on an AST?

Too complicated. @felixhaeberle just worked with JSONC in VSCode. It's horror.

**Comment by openscript** [comment]
Will it be possible to provide a local plugin to the config or can inlangs dynamic behavior only be customized via external plugins afters this change?

**Comment by samuelstroschein** [comment]
> Will it be possible to provide a local plugin to the config or can inlangs dynamic behavior only be customized via external plugins afters this change? 

Local plugins will be supported. The module link is a local path in that case.

**Comment by samuelstroschein** [comment]
Closing as "decided to have JSON as config format"
